homeassistant/components/upnp/config_flow.py

```diff
@@ -1,10 +1,172 @@
+        ]
+        self._discoveries = [
+            discovery
+            for discovery in self._discoveries
+            if discovery["unique_id"] not in current_unique_ids
+        ]
+
+        # Ensure anything to add.
+        if not self._discoveries:
+            errors = {"base": "no_devices_discovered"}
+            return self.async_show_form(step_id="user", errors=errors,)
+
+        data_schema = vol.Schema(
+            {
+                vol.Required("unique_id"): vol.In(
+                    {
+                        discovery["unique_id"]: urlparse(
+                            discovery[ssdp.ATTR_SSDP_LOCATION]
+                        ).hostname
+                        for discovery in self._discoveries
+                    }
+                ),
+            }
+        )
+        return self.async_show_form(step_id="user", data_schema=data_schema,)
+
+    async def async_step_import(self, import_info: Optional[Mapping]):
+        """Import a new UPnP/IGD device as a config entry.
+
+        This flow is triggered by `async_setup`.
+        """
+        _LOGGER.debug("async_step_import: import_info: %s", import_info)
+
+        if import_info is None:
+            # Landed here via configuration.yaml entry.
+            # Any device already added, then abort.
+            if self._async_current_entries():
+                _LOGGER.debug("aborting, already configured")
+                return self.async_abort(reason="already_configured")
+
+        # Test if import_info isn't already configured.
+        if import_info is not None and any(
+            import_info["udn"] == entry.data["udn"]
+            and import_info["st"] == entry.data["st"]
+            for entry in self._async_current_entries()
+        ):
+            return self.async_abort(reason="already_configured")
+
+        # Discover devices.
+        self._discoveries = await Device.async_discover(self.hass)
+
+        # Ensure anything to add. If not, silently abort.
+        if not self._discoveries:
+            _LOGGER.info("No UPnP devices discovered, aborting.")
+            return self.async_abort(reason="no_devices_found")
+
+        # Create new config_entry.
+        self._data = self._discoveries[0]
+        return self._async_create_entry_from_data()
+
+    async def async_step_ssdp(self, discovery_info: Mapping):
+        """Handle a discovered UPnP/IGD device.
+
+        This flow is triggered by the SSDP component. It will check if the
+        host is already configured and delegate to the import step if not.
+        """
+        _LOGGER.debug("async_step_ssdp: discovery_info: %s", discovery_info)
+
+        # Ensure not already configuring/configured.
+        udn = discovery_info[ssdp.ATTR_UPNP_UDN]
+        st = discovery_info[ssdp.ATTR_SSDP_ST]  # pylint: disable=invalid-name
+        usn = f"{udn}::{st}"
+        await self.async_set_unique_id(usn)
+        self._abort_if_unique_id_configured(updates={UDN: udn, ST: st})
```

Comment:
Since unique ID is based on `usn`, you will never encounter a necessary update, because there wouldn't be a match for UDN or ST if the unique ID matche.s